File: binary_search.py
def binary_search(array, value):
    """Search for a value in an array using binary search.

    Return index of value if found; -1 otherwise
    Time complexity: O(logn)
    Space complexity: O(1)
    """
    low = 0
    high = len(array)-1
    while low <= high:
        mid = (low+high)/2
        if array[mid] == value:
            return mid
        elif array[mid] < value:
            low = mid + 1
        elif array[mid] > value:
            high = mid - 1
    return -1

array = [1,3,9,11,15,19,29]
value1 = 25
value2 = 15
print binary_search(array, value1)
print binary_search(array, value2)

File: java/BinarySearch.java
public class BinarySearch {

    /**
     * This method finds an integer n using binary search.
     *
     * <p>Time complexity: O(logn) | Space complexity: O(1)
     *
     * @param arr the array of integers
     * @param n   the integer to find in arr
     * @return    true if n is found in arr; false otherwise
     */
    public static boolean binarySearch(int[] arr, int n) {
        int low = 0;
        int high = arr.length - 1;
        while (high >= low) {
            int mid = (low + high) / 2;
            if (n == arr[mid]) {
                return true;
            } else if (n > arr[mid]) {
                low = mid + 1;
            } else if (n < arr[mid]) {
                high = mid - 1;
            }
        }
        return false;
    }

    public static void main(String[] args) {
        System.out.println(binarySearch(new int[] {2,5,7,8,9,12}, 8)); // true
        System.out.println(binarySearch(new int[] {2}, 4)); // false
        System.out.println(binarySearch(new int[] {2}, 2)); // true
        System.out.println(binarySearch(new int[] {2,5,7,9,12}, 8)); // false
        System.out.println(binarySearch(new int[] {2,5,7,9,12}, 12)); // true
        System.out.println(binarySearch(new int[] {2,5,7,9,12}, 2)); // true
        System.out.println(binarySearch(new int[] {}, 4)); // false
    }
}
